File: modal-js/test/sandbox_filesystem_snapshot.test.ts
import { tc } from "../test-support/test-client";
import { expect, test, onTestFinished } from "vitest";

test("snapshotFilesystem", async () => {
  const app = await tc.apps.fromName("libmodal-test", {
    createIfMissing: true,
  });
  const image = tc.images.fromRegistry("alpine:3.21");

  const sb = await tc.sandboxes.create(app, image);
  onTestFinished(async () => {
    await sb.terminate();
  });

  await sb.exec(["sh", "-c", "echo -n 'test content' > /tmp/test.txt"]);
  await sb.exec(["mkdir", "-p", "/tmp/testdir"]);

  const snapshotImage = await sb.snapshotFilesystem();
  expect(snapshotImage).toBeDefined();
  expect(snapshotImage.imageId).toMatch(/^im-/);

  await sb.terminate();

  // Create new sandbox from snapshot
  const sb2 = await tc.sandboxes.create(app, snapshotImage);
  onTestFinished(async () => {
    await sb2.terminate();
  });

  // Verify file exists in snapshot
  const proc = await sb2.exec(["cat", "/tmp/test.txt"]);
  const output = await proc.stdout.readText();
  expect(output).toBe("test content");

  // Verify directory exists in snapshot
  const dirCheck = await sb2.exec(["test", "-d", "/tmp/testdir"]);
  expect(await dirCheck.wait()).toBe(0);
});

File: modal-go/test/sandbox_filesystem_snapshot_test.go
package test

import (
	"context"
	"io"
	"testing"
	"time"

	"github.com/modal-labs/libmodal/modal-go"
	"github.com/onsi/gomega"
)

func TestSnapshotFilesystem(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := context.Background()
	tc := newTestClient(t)

	app, err := tc.Apps.FromName(ctx, "libmodal-test", &modal.AppFromNameParams{CreateIfMissing: true})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	image := tc.Images.FromRegistry("alpine:3.21", nil)

	sb, err := tc.Sandboxes.Create(ctx, app, image, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer terminateSandbox(g, sb)

	_, err = sb.Exec(ctx, []string{"sh", "-c", "echo -n 'test content' > /tmp/test.txt"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	_, err = sb.Exec(ctx, []string{"mkdir", "-p", "/tmp/testdir"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	snapshotImage, err := sb.SnapshotFilesystem(ctx, 55*time.Second)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(snapshotImage).ShouldNot(gomega.BeNil())
	g.Expect(snapshotImage.ImageID).To(gomega.HavePrefix("im-"))

	err = sb.Terminate(ctx, true, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	sb2, err := tc.Sandboxes.Create(ctx, app, snapshotImage, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer terminateSandbox(g, sb2)

	proc, err := sb2.Exec(ctx, []string{"cat", "/tmp/test.txt"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	output, err := io.ReadAll(proc.Stdout)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(string(output)).To(gomega.Equal("test content"))

	dirCheck, err := sb2.Exec(ctx, []string{"test", "-d", "/tmp/testdir"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	exitCode, err := dirCheck.Wait(ctx)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(exitCode).To(gomega.Equal(0))
}
